android-beacon/src/main/java/com/novoda/beacon/MessageService.java

```diff
@@ -0,0 +1,90 @@
+package com.novoda.beacon;
+
+import android.app.PendingIntent;
+import android.app.Service;
+import android.content.Intent;
+import android.os.Bundle;
+import android.os.IBinder;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.util.Log;
+
+import com.google.android.gms.common.ConnectionResult;
+import com.google.android.gms.common.api.GoogleApiClient;
+import com.google.android.gms.nearby.Nearby;
+import com.google.android.gms.nearby.messages.Strategy;
+import com.google.android.gms.nearby.messages.SubscribeOptions;
+
+public class MessageService extends Service {
+
+    private static final String TAG = "!!!";
+
+    private GoogleApiClient googleApiClient;
+
+    @Nullable
+    @Override
+    public IBinder onBind(Intent intent) {
+        return null;
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        googleApiClient = new GoogleApiClient.Builder(this)
+                .addApi(Nearby.MESSAGES_API)
+                .addConnectionCallbacks(connectionCallbacks)
+                .addOnConnectionFailedListener(onConnectionFailedListener)
+                .build();
+
+        googleApiClient.connect();
+        return Service.START_NOT_STICKY;
+    }
+
+    private final GoogleApiClient.ConnectionCallbacks connectionCallbacks = new GoogleApiClient.ConnectionCallbacks() {
+        @Override
+        public void onConnected(@Nullable Bundle bundle) {
+            Log.d(TAG, "onConnectionConnected");
+            subscribe(googleApiClient);
+        }
+
+        private void subscribe(GoogleApiClient googleApiClient) {
+            Log.i(TAG, "Subscribing.");
+            SubscribeOptions options = new SubscribeOptions.Builder()
+                    .setStrategy(Strategy.BLE_ONLY)
+                    .build();
+            Nearby.Messages.subscribe(googleApiClient, createPendingIntent(), options);
+        }
+
+        private PendingIntent createPendingIntent() {
+            Intent receiver = new Intent(MessageService.this, MessageReceiver.class);
+            return PendingIntent.getBroadcast(MessageService.this, 0, receiver, PendingIntent.FLAG_UPDATE_CURRENT);
+        }
+
+        @Override
+        public void onConnectionSuspended(int i) {
+            Log.d(TAG, "onConnectionSuspended");
+        }
+    };
+
+    private final GoogleApiClient.OnConnectionFailedListener onConnectionFailedListener = new GoogleApiClient.OnConnectionFailedListener() {
+        @Override
+        public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
+            Log.d(TAG, "onConnectionFailed");
+            attemptToResolve(connectionResult);
+        }
+
+        private void attemptToResolve(@NonNull ConnectionResult connectionResult) {
+            try {
+                connectionResult.getResolution().send(MessageService.this, MainActivity.RESOLVE_PERMISSIONS, null);
```

Comment:
So, basically when a `ConnectionResult` fails, it can be resolved by the user (for example when missing a permission, or requiring Play Services update).
`getResolution()` returns a PendingIntent to the activity responsible of the error resolution: the Play Store for an update, a dialog for the notification to be displayed to the user, etc.
So calling `that resolution PendingIntent.send()` is gonna open an activity for result with the Result code we pass, and the result of that activity is caught by our activity, which restarts the service